banzai_nres/rv.py

```diff
@@ -0,0 +1,133 @@
+from banzai.stages import Stage
+from banzai.frames import ObservationFrame
+from banzai import dbs
+import pkg_resources
+from astropy.io import fits
+import numpy as np
+from astropy.table import Table
+from astropy import constants
+from astropy.time import Time
+from astropy.coordinates import SkyCoord, EarthLocation, solar_system_ephemeris
+from astropy import units
+from banzai.utils import stats
+import logging
+
+
+logger = logging.getLogger('banzai')
+
+# Speed of light in km/s
+c = constants.c.to('km / s').value
+
+
+def cross_correlate(velocities, wavelength, flux, flux_uncertainty, template_wavelength, template_flux):
+    """
+    :param velocities: in km / s
+    :param wavelength:
+    :param flux:
+    :param flux_uncertainty:
+    :param template_wavelength:
+    :param template_flux:
+    :return:
+
+    This is consistent with what we did in the satellite pipeline and
+    Zackay & Ofek, 2017, ApJ, 836, 187
+    """
+    x_cor = []
+    # for steps in 1 km/s from -2000 to +2000 km/s
+    # calculate the variance ahead of time
+    variance = flux_uncertainty * flux_uncertainty
+
+    for v in velocities:
+        doppler = 1.0 / (1.0 + v / c)
+        kernel = np.interp(doppler * wavelength, template_wavelength, template_flux)
+        correlation = kernel * flux
+        correlation /= variance
+
+        normalization = kernel * kernel / variance
+
+        # This gives us the signal to noise ratio in terms of N-sigma
+        snr = correlation.sum() / np.sqrt(normalization.sum())
+        x_cor.append(snr)
+
+    return np.array(x_cor)
+
+
+def barycentric_correction(time, exptime, ra, dec, site):
+    # The following is largely taken from the astropy example page
+    site_location = EarthLocation.from_geodetic(lat=site.latitude*units.deg, lon=site.longitude*units.deg,
+                                                height=site.elevation*units.m)
+    sky_coordinates = SkyCoord(ra=ra, dec=dec, unit=(units.hourangle, units.deg))
+    # The time given in the NRES header is the exposure start time; correct this to the midpoint
+    # Eventually, we will want to use the flux-weighted mid-point calculated from emeter data
+    obs_time = Time(time, location=site_location) + exptime / 2.0 * units.s
+    barycorr_rv = sky_coordinates.radial_velocity_correction(obstime=obs_time)
+    bjd_tdb = obs_time.tdb + obs_time.light_travel_time(sky_coordinates)
+    return barycorr_rv.to('m/s').value, bjd_tdb.to_value('jd')
```

Comment:
This returns the time in JD, not BJD (I think). Yet Comment 2 below thinks it is a BJD.